### ISSUE TITLE
Feature Request: Guard against pushing private pods to trunk.

### ISSUE THREAD

**Issue by brianpartridge** [CocoaPods/CocoaPods#3275]
Within my team there is concern that we could accidentally publish a podspec for an internal library to trunk rather than our private pods repo.  Since we have public and private pods, it's possible that a developer could accidentally run the wrong push command when publishing a new version.  I'm proposing an optional setting in `Specification` which would guard against this:

- Add a new field to `Specification` named `internal`/`private`/`can_publish_to_trunk`.
- Detect this flag in `pod trunk push` and fail accordingly.

Does this sound reasonable? All specs default to supporting public publishing, but teams that care about this could ensure their internal podspec template disables this.

I'm not asking the CocoaPods team to implement this feature (I'll do it), I just want to see if it would be considered acceptable before starting on it. Thanks!

**Comment by neonichu** [comment]
You can get a similarish behaviour by using the recently introduced `COCOAPODS_TRUNK_TOKEN` environment variable (see #3224). If you keep that set to something bogus by default (via some shell configuration file), it would prevent any pushes to trunk while manually unsetting it would allow it to work.

```
$ export COCOAPODS_TRUNK_TOKEN=''
$ pod trunk push # => fail
$ unset COCOAPODS_TRUNK_TOKEN
$ pod trunk push # => success
```

**Comment by segiddins** [comment]
I don't think this is really necessary -- your specification won't ever get pushed to trunk unless you push it there.

**Comment by brianpartridge** [comment]
The environment variable may work for us.

@segiddins the concern is that there is only a slight difference in workflow when pushing to trunk versus pushing to a private spec repo.  If we had to *add* a project to trunk before pushing the team would not have a concern. As it stands, it's too easy to hit up in the terminal a couple times see the word "push" and then just change the podspec name.

**Comment by orta** [comment]
Given the amount of times this has happened, and the difficulty in general in removing something in contrast to sending it, I think it might make sense to expand the podspec DSL for something like this.

**Comment by brianpartridge** [comment]
Thanks @orta I'll try to PR something later this week.

**Comment by kylef** [comment]
I don't think I like the approach of adding extra flags to the DSL for this. It should be necessary.

Instead, I think it would make more sense to fix the underlying problem, that `pod trunk` uses users credentials inside `.netrc` to download the source of a public pod. Since, any Pod from the master spec repository should be downloadable without any credentials, we shouldn't lint it with the users credentials. Preventing users pushing private pods to the public spec repository.

Fixing this problem would ultimately prevent you pushing a pod accidentally to trunk because it would fail linting if the source was private.

**Comment by brianpartridge** [comment]
@kylef I mostly agree, it would be best if trunk could validate pushes based on the source code availability, but how would you characterize if the source is private?  It can't just rely upon the source being hosted on github because repos there can be private, same with BitBucket, and other third parties.

**Comment by kylef** [comment]
@brianpartridge I don't think I follow your question. If trunk can download the source without authentication, then it is deemed public.

**Comment by orta** [comment]
I think the gist is that we can add support to `trunk.cocoapods.org`, which has none of the user's authentication systems , to check to see if it can see the private repo. If it cannot access it, then it shouldn't be on the public repo.

**Comment by brianpartridge** [comment]
Ah, I see. I thought @kylef meant the `trunk` command rather than the server.  Yes, that would be fine by me.

**Comment by brianpartridge** [comment]
Alright, I'm going to nuke my PRs. I'd like to make a formal feature request to validate 'public' specs on the trunk service.  Thanks for the discussion.

Should I close this and open another issue? If so, here or on the trunk repo?

**Comment by orta** [comment]
Yeah, lets get it in a single fresh issue :👍
